src/NuGet.Core/NuGet.Protocol/Resources/RawSearchResourceV3.cs

```diff
@@ -103,22 +108,84 @@ public virtual async Task<JObject> SearchPage(string searchTerm, SearchFilter fi
                     throw new FatalProtocolException(string.Format(CultureInfo.CurrentCulture, Strings.Protocol_BadSource, queryUrl.Uri), ex);
                 }
 
-                if (searchJson != null)
+                if (searchResult != null)
                 {
-                    return searchJson;
+                    return searchResult;
                 }
             }
 
             // TODO: get a better message for this
             throw new FatalProtocolException(Strings.Protocol_MissingSearchService);
         }
 
-        public virtual async Task<IEnumerable<JObject>> Search(string searchTerm, SearchFilter filters, int skip, int take, Common.ILogger log, CancellationToken cancellationToken)
+        private async Task<T> Search<T>(
+            Func<HttpSource, Uri, Task<T>> getResultAsync,
+            string searchTerm,
+            SearchFilter filters,
+            int skip,
+            int take,
+            Common.ILogger log,
+            CancellationToken cancellationToken)
         {
-            var results = await SearchPage(searchTerm, filters, skip, take, log, cancellationToken);
+            return await SearchPage(
+                uri => getResultAsync(_client, uri),
+                searchTerm,
+                filters,
+                skip,
+                take,
+                log,
+                cancellationToken);
+        }
+
+        internal async Task<IEnumerable<PackageSearchMetadata>> Search(
```

Comment:
I'd add docs here to clarify that we're stopping early.

Reply:
Same applies to the PackageSearchResourceV3.

Reply:
Ok, added comment summary on each new API. Please review.